src/main/java/com/blog/stream/pagination/PageSpliterator.java

```diff
@@ -1,108 +1,189 @@
 package com.blog.stream.pagination;
 
-import com.google.common.annotations.VisibleForTesting;
-import org.springframework.data.domain.Page;
-
+import java.util.List;
 import java.util.Spliterator;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiFunction;
 import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * spliterator can iterate through page results for all pages sequentially,
+ * or split on pages to allow fetching in parallel, but will get an accurate page count by prefetching
+ *
+ * @author ed.wenwtworth - inspired by/forked from https://github.com/mattsibs/pagination-stream
+ * @param <P> page object returned
+ * @param <T> type of object
+ */
+public class PageSpliterator<P, T> implements Spliterator<T> {
```

Comment:
Think rather than having P, maybe worth just creating a simple DTO to replace the JPA page request type?